metals/src/main/scala/scala/meta/internal/metals/debug/SourcePathAdapter.scala

```diff
@@ -17,12 +19,36 @@ private[debug] final class SourcePathAdapter(
     workspace: AbsolutePath,
     sources: Set[AbsolutePath],
     buildTargets: BuildTargets,
-    saveJarFileToDisk: Boolean
+    supportVirtualDocuments: Boolean
 ) {
+  // when virtual documents are supported there is no need to save jars on disk
+  private val saveJarFileToDisk = !supportVirtualDocuments
   private val dependencies = workspace.resolve(Directories.dependencies)
+
   def toDapURI(sourcePath: AbsolutePath): Option[URI] = {
     if (sources.contains(sourcePath)) Some(sourcePath.toURI)
-    else {
+    else if (supportVirtualDocuments) {
+      val sourceFileJarPath = workspace.toNIO
+        .relativize(sourcePath.toNIO)
+        .toString
+        .stripPrefix("jar:")
+      val decodedPath =
+        URIEncoderDecoder.decode(sourceFileJarPath).toAbsolutePath
+      val parts = decodedPath.toNIO.iterator().asScala.map(_.toString).toVector
+      val jarPartIndex = parts.indexWhere(path => path.endsWith(".jar!"))
+      val jarPath = AbsolutePath(
+        parts
+          .take(jarPartIndex + 1)
+          .mkString(File.separator, File.separator, "")
+          .stripSuffix("!")
+      )
+      val relative = parts.drop(jarPartIndex + 1).mkString(File.separator)
```

Comment:
Honestly, this code is disgusting, but I don't have any other idea how to turn path like
```scala
AbsolutePath(
  toNIO = /Users/kpodsiadlo/vl/my-app/jar:file%3A/Users/kpodsiadlo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.8/scala-library-2.13.8-sources.jar%21/scala/collection/immutable/List.scala
)
```
into base jar path + relative counterpart

Reply:
I'm hoping to get rid of the jar encoding when I finally get the filesystem api working so I wouldn't spend much time on making it pretty.

Reply:
As long as it works it's fine, it's not that bad :sweat_smile:

Reply:
Ok, so if it's not that bad and it's a huge chance that it's a temporary workaround I can leave it as it is now.

Reply:
I think the bizarre relative path format is down to VSCode only supporting `pathFormat` of `path`, not `uri`.  So all non `file:///` uris are just prepended with the workspace path.  There's an issue [here](https://github.com/microsoft/vscode/issues/68808)